### ISSUE TITLE
vcferr abandons phase

### ISSUE THREAD

**Issue by stephenturner** [signaturescience/vcferr#20]
VCFerr abandons phase information. It may be desirable to retain phase information, (e.g., when using `--keep_phase` with something like ped-sim, or when using already phased data). 

Demonstration / reprex:

Look at the example data:

```
$ bcftools view -H data/example.vcf.gz
1	1	.	A	C	.	.	.	GT	0/0	0/0	1/1
1	2	.	G	A	.	.	.	GT	0/0	0/0	1/1
1	3	.	T	T	.	.	.	GT	0/1	0/0	1/1
1	4	.	C	T	.	.	.	GT	0/1	0/0	1/1
1	5	.	T	T	.	.	.	GT	0/1	0/0	1/1
1	6	.	T	G	.	.	.	GT	0/1	0/0	1/1
1	7	.	G	C	.	.	.	GT	1/1	0/0	1/1
1	8	.	C	A	.	.	.	GT	1/1	0/0	1/1
1	9	.	A	T	.	.	.	GT	1/1	0/0	1/1
2	1	.	C	A	.	.	.	GT	0/0	0/0	1/1
2	2	.	C	G	.	.	.	GT	0/0	0/0	1/1
2	3	.	C	T	.	.	.	GT	0/1	0/0	1/1
2	4	.	A	G	.	.	.	GT	0/1	0/0	1/1
2	5	.	G	T	.	.	.	GT	0/1	0/0	1/1
2	6	.	A	C	.	.	.	GT	0/1	0/0	1/1
2	7	.	A	A	.	.	.	GT	1/1	0/0	1/1
2	8	.	T	A	.	.	.	GT	1/1	0/0	1/1
2	9	.	A	G	.	.	.	GT	1/1	0/0	1/1
```

Set all the genotypes to phased:

```
$ bcftools +setGT data/example.vcf.gz -- -t a -n p | bcftools view -Oz -o data/phased.vcf.gz && tabix -f data/phased.vcf.gz
Filled 108 alleles
$ bcftools view -H data/phased.vcf.gz
1	1	.	A	C	.	.	.	GT	0|0	0|0	1|1
1	2	.	G	A	.	.	.	GT	0|0	0|0	1|1
1	3	.	T	T	.	.	.	GT	0|1	0|0	1|1
1	4	.	C	T	.	.	.	GT	0|1	0|0	1|1
1	5	.	T	T	.	.	.	GT	0|1	0|0	1|1
1	6	.	T	G	.	.	.	GT	0|1	0|0	1|1
1	7	.	G	C	.	.	.	GT	1|1	0|0	1|1
1	8	.	C	A	.	.	.	GT	1|1	0|0	1|1
1	9	.	A	T	.	.	.	GT	1|1	0|0	1|1
2	1	.	C	A	.	.	.	GT	0|0	0|0	1|1
2	2	.	C	G	.	.	.	GT	0|0	0|0	1|1
2	3	.	C	T	.	.	.	GT	0|1	0|0	1|1
2	4	.	A	G	.	.	.	GT	0|1	0|0	1|1
2	5	.	G	T	.	.	.	GT	0|1	0|0	1|1
2	6	.	A	C	.	.	.	GT	0|1	0|0	1|1
2	7	.	A	A	.	.	.	GT	1|1	0|0	1|1
2	8	.	T	A	.	.	.	GT	1|1	0|0	1|1
2	9	.	A	G	.	.	.	GT	1|1	0|0	1|1
```

Run vcferr on sample 1 and observe that that sample's data is no longer phased:

```
$ vcferr data/phased.vcf.gz --sample='sample1' --p_rarr=.5 | bcftools view -H
1	1	.	A	C	.	.	.	GT	0/0	0|0	1|1
1	2	.	G	A	.	.	.	GT	0/0	0|0	1|1
1	3	.	T	T	.	.	.	GT	0/0	0|0	1|1
1	4	.	C	T	.	.	.	GT	0/0	0|0	1|1
1	5	.	T	T	.	.	.	GT	0/1	0|0	1|1
1	6	.	T	G	.	.	.	GT	0/1	0|0	1|1
1	7	.	G	C	.	.	.	GT	1/1	0|0	1|1
1	8	.	C	A	.	.	.	GT	1/1	0|0	1|1
1	9	.	A	T	.	.	.	GT	1/1	0|0	1|1
2	1	.	C	A	.	.	.	GT	0/0	0|0	1|1
2	2	.	C	G	.	.	.	GT	0/0	0|0	1|1
2	3	.	C	T	.	.	.	GT	0/0	0|0	1|1
2	4	.	A	G	.	.	.	GT	0/1	0|0	1|1
2	5	.	G	T	.	.	.	GT	0/1	0|0	1|1
2	6	.	A	C	.	.	.	GT	0/1	0|0	1|1
2	7	.	A	A	.	.	.	GT	1/1	0|0	1|1
2	8	.	T	A	.	.	.	GT	1/1	0|0	1|1
2	9	.	A	G	.	.	.	GT	1/1	0|0	1|1
```

If doing this with simulated data where all phase information is preserved, we can simply use the same bcftools trick to re-phase. But if there's a built-in way to maintain phase, that would be preferable.

**Comment by vpnagraj** [comment]
this might be very straightforward to add:

https://github.com/pysam-developers/pysam/issues/849#issuecomment-540622581

more soon ...

**Comment by vpnagraj** [comment]
@stephenturner i think i have a fix for this at https://github.com/signaturescience/vcferr/tree/keep-phase

switch to that branch and `pip install .`

using the example above:

```
vcferr data/phased.vcf.gz --sample='sample1' --p_rarr=.5 --phased False | bcftools view -H
```

```
1       1       .       A       C       .       .       .       GT      0/0     0|0     1|1
1       2       .       G       A       .       .       .       GT      0/0     0|0     1|1
1       3       .       T       T       .       .       .       GT      0/1     0|0     1|1
1       4       .       C       T       .       .       .       GT      0/1     0|0     1|1
1       5       .       T       T       .       .       .       GT      0/0     0|0     1|1
1       6       .       T       G       .       .       .       GT      0/0     0|0     1|1
1       7       .       G       C       .       .       .       GT      1/1     0|0     1|1
1       8       .       C       A       .       .       .       GT      1/1     0|0     1|1
1       9       .       A       T       .       .       .       GT      1/1     0|0     1|1
2       1       .       C       A       .       .       .       GT      0/0     0|0     1|1
2       2       .       C       G       .       .       .       GT      0/0     0|0     1|1
2       3       .       C       T       .       .       .       GT      0/1     0|0     1|1
2       4       .       A       G       .       .       .       GT      0/1     0|0     1|1
2       5       .       G       T       .       .       .       GT      0/1     0|0     1|1
2       6       .       A       C       .       .       .       GT      0/0     0|0     1|1
2       7       .       A       A       .       .       .       GT      1/1     0|0     1|1
2       8       .       T       A       .       .       .       GT      1/1     0|0     1|1
2       9       .       A       G       .       .       .       GT      1/1     0|0     1|1
```

```
vcferr data/phased.vcf.gz --sample='sample1' --p_rarr=.5 --phased True | bcftools view -H 
```

```
1       1       .       A       C       .       .       .       GT      0|0     0|0     1|1
1       2       .       G       A       .       .       .       GT      0|0     0|0     1|1
1       3       .       T       T       .       .       .       GT      0|0     0|0     1|1
1       4       .       C       T       .       .       .       GT      0|1     0|0     1|1
1       5       .       T       T       .       .       .       GT      0|1     0|0     1|1
1       6       .       T       G       .       .       .       GT      0|0     0|0     1|1
1       7       .       G       C       .       .       .       GT      1|1     0|0     1|1
1       8       .       C       A       .       .       .       GT      1|1     0|0     1|1
1       9       .       A       T       .       .       .       GT      1|1     0|0     1|1
2       1       .       C       A       .       .       .       GT      0|0     0|0     1|1
2       2       .       C       G       .       .       .       GT      0|0     0|0     1|1
2       3       .       C       T       .       .       .       GT      0|1     0|0     1|1
2       4       .       A       G       .       .       .       GT      0|0     0|0     1|1
2       5       .       G       T       .       .       .       GT      0|1     0|0     1|1
2       6       .       A       C       .       .       .       GT      0|1     0|0     1|1
2       7       .       A       A       .       .       .       GT      1|1     0|0     1|1
2       8       .       T       A       .       .       .       GT      1|1     0|0     1|1
2       9       .       A       G       .       .       .       GT      1|1     0|0     1|1
```

give it a shot. if you have other data to test against that would be great too. if it looks good go ahead and close the issue, and i'll bump the version and pull to main.

**Comment by stephenturner** [comment]
I think this is good enough. My original thought was to maintain phase. I.e., if it was phased, keep it phased, if it wasn't phased, keep it unphased. Here's an example of erroneous usage. It takes the unphased input, introduces errors, and gives you _phased output_ just for that sample. 

```
vcferr data/example.vcf.gz --sample='sample1' --p_rarr=.5 --phased True | bcftools view -H
1	1	.	A	C	.	.	.	GT	0|0	0/0	1/1
1	2	.	G	A	.	.	.	GT	0|0	0/0	1/1
1	3	.	T	T	.	.	.	GT	0|1	0/0	1/1
1	4	.	C	T	.	.	.	GT	0|1	0/0	1/1
1	5	.	T	T	.	.	.	GT	0|1	0/0	1/1
```

But I didn't think this through carefully. There are a couple options I see:

1. Keep it like it's written now. User would have to know whether input genotypes are phased or not, and remember to use the correct option to `--phased`. 
2. Provide an option that _maintains_ phase. You'd first have to check whether `rec.samples[sample]['GT']` is phased, and if so, return phased output.
3. Similar to 2 above, but don't provide an option. Just always maintain phase. I can't imagine a scenario where we'd want something different. 

Advantages to _maintaining_ phase instead of _setting_ phase, especially if it's not an option, besides not having to remember whether to turn it on or not, might be the case where some parts of the chromosome are phased and others aren't (won't happen in simulation but might in real life, depending on how data was phased?). If this is happening on a record by record level, seems like this might be possible?

I can generate some test data for that case if needed.

**Comment by vpnagraj** [comment]
good call. option 3 is best IMO.

ive updated the code on the `keep-phase` branch to handle this:

```
vcferr data/phased.vcf.gz --sample='sample1' --p_rarr=.5 | bcftools view -H
```

```
1       1       .       A       C       .       .       .       GT      0|0     0|0     1|1
1       2       .       G       A       .       .       .       GT      0|0     0|0     1|1
1       3       .       T       T       .       .       .       GT      0|1     0|0     1|1
1       4       .       C       T       .       .       .       GT      0|0     0|0     1|1
1       5       .       T       T       .       .       .       GT      0|0     0|0     1|1
1       6       .       T       G       .       .       .       GT      0|0     0|0     1|1
1       7       .       G       C       .       .       .       GT      1|1     0|0     1|1
1       8       .       C       A       .       .       .       GT      1|1     0|0     1|1
1       9       .       A       T       .       .       .       GT      1|1     0|0     1|1
2       1       .       C       A       .       .       .       GT      0|0     0|0     1|1
2       2       .       C       G       .       .       .       GT      0|0     0|0     1|1
2       3       .       C       T       .       .       .       GT      0|0     0|0     1|1
2       4       .       A       G       .       .       .       GT      0|1     0|0     1|1
2       5       .       G       T       .       .       .       GT      0|1     0|0     1|1
2       6       .       A       C       .       .       .       GT      0|0     0|0     1|1
2       7       .       A       A       .       .       .       GT      1|1     0|0     1|1
2       8       .       T       A       .       .       .       GT      1|1     0|0     1|1
2       9       .       A       G       .       .       .       GT      1|1     0|0     1|1
```


```
vcferr data/example.vcf.gz --sample='sample1' --p_rarr=.5 | bcftools view -H
```


```
1       1       .       A       C       .       .       .       GT      0/0     0/0     1/1
1       2       .       G       A       .       .       .       GT      0/0     0/0     1/1
1       3       .       T       T       .       .       .       GT      0/1     0/0     1/1
1       4       .       C       T       .       .       .       GT      0/1     0/0     1/1
1       5       .       T       T       .       .       .       GT      0/0     0/0     1/1
1       6       .       T       G       .       .       .       GT      0/0     0/0     1/1
1       7       .       G       C       .       .       .       GT      1/1     0/0     1/1
1       8       .       C       A       .       .       .       GT      1/1     0/0     1/1
1       9       .       A       T       .       .       .       GT      1/1     0/0     1/1
2       1       .       C       A       .       .       .       GT      0/0     0/0     1/1
2       2       .       C       G       .       .       .       GT      0/0     0/0     1/1
2       3       .       C       T       .       .       .       GT      0/1     0/0     1/1
2       4       .       A       G       .       .       .       GT      0/1     0/0     1/1
2       5       .       G       T       .       .       .       GT      0/1     0/0     1/1
2       6       .       A       C       .       .       .       GT      0/1     0/0     1/1
2       7       .       A       A       .       .       .       GT      1/1     0/0     1/1
2       8       .       T       A       .       .       .       GT      1/1     0/0     1/1
2       9       .       A       G       .       .       .       GT      1/1     0/0     1/1
```

**Comment by stephenturner** [comment]
Confirmed this works, and with partially phased genotypes also. Code below makes a mock sample where the 0/0 genotypes are set to phased only for sample 1.

```
bcftools view data/example.vcf.gz | sed 's@0/0@0|0@' | bcftools view -Oz -o data/partiallyphased.vcf.gz
tabix data/partiallyphased.vcf.gz

vcferr --sample 'sample1' -p_rraa 1 data/partiallyphased.vcf.gz | bcftools view -H
1	1	.	A	C	.	.	.	GT	1|1	0/0	1/1
1	2	.	G	A	.	.	.	GT	1|1	0/0	1/1
1	3	.	T	T	.	.	.	GT	0/1	0|0	1/1
1	4	.	C	T	.	.	.	GT	0/1	0|0	1/1
1	5	.	T	T	.	.	.	GT	0/1	0|0	1/1
1	6	.	T	G	.	.	.	GT	0/1	0|0	1/1
1	7	.	G	C	.	.	.	GT	1/1	0|0	1/1
1	8	.	C	A	.	.	.	GT	1/1	0|0	1/1
1	9	.	A	T	.	.	.	GT	1/1	0|0	1/1
2	1	.	C	A	.	.	.	GT	1|1	0/0	1/1
2	2	.	C	G	.	.	.	GT	1|1	0/0	1/1
2	3	.	C	T	.	.	.	GT	0/1	0|0	1/1
2	4	.	A	G	.	.	.	GT	0/1	0|0	1/1
2	5	.	G	T	.	.	.	GT	0/1	0|0	1/1
2	6	.	A	C	.	.	.	GT	0/1	0|0	1/1
2	7	.	A	A	.	.	.	GT	1/1	0|0	1/1
2	8	.	T	A	.	.	.	GT	1/1	0|0	1/1
2	9	.	A	G	.	.	.	GT	1/1	0|0	1/1
```

Go ahead and open a PR / version bump.

**Comment by vpnagraj** [comment]
resolved in #21
